File: AlgorithmDemo/QuickSort.c
//
//  QuickSort.c
//  AlgorithmDemo
//
//  Created by 刘宇 on 2017/8/30.
//  Copyright © 2017年 刘宇. All rights reserved.
//

#include "QuickSort.h"

void swap(int *i, int *j)
{
    int tmp = *i;
    *i = *j;
    *j = tmp;
}

void quickSort(int *array , int low, int height)
{
    if (low >= height) {
        return;
    }
    int i = low;
    int j = height;
    int key = array[i];
    while (i < j)
    {
        while (i < j && array[j] >= key)
        {
            j--;
        }
        swap(&array[i],&array[j]);

        while (i < j && array[i] <= key) {
            i++;
        }
        swap(&array[i],&array[j]);

    }
    quickSort(array, low, i - 1);
    quickSort(array, i + 1, height);
}

void quickSort2(int array[], int low, int height)
{
    if (low >= height) {
        return;
    }
    int i = low;
    int j = height;
    int key = array[i];
    while (i < j)
    {
        while (i < j && array[j] >= key)
        {
            j--;
        }
        array[i] = array[j];

        while (i < j && array[i] <= key) {
            i++;
        }
        array[j] = array[i];

    }
    array[i] = key;
    quickSort(array, low, i - 1);
    quickSort(array, i + 1, height);
}

File: OC_algorithmDemo/Classes/QuickSort.m
//
//  QuickSort.m
//  OC_algorithmDemo
//
//  Created by 刘宇 on 2017/8/31.
//  Copyright © 2017年 刘宇. All rights reserved.
//

#import "QuickSort.h"

@implementation QuickSort

- (NSArray<NSNumber *>*)startRun
{
   return [self sortWithArray:[NSMutableArray arrayWithArray:self.array] leftIndex:0 rightIndex:self.array.count - 1];
}

- (NSArray<NSNumber *>*)sortWithArray:(NSMutableArray<NSNumber*>*)array leftIndex:(NSInteger)leftIndex rightIndex:(NSInteger)rightIndex
{
    if (leftIndex > rightIndex) {
        return array;
    }
    
    NSInteger key = array[leftIndex].integerValue;
    NSInteger i = leftIndex;
    NSInteger j = rightIndex;
    
    while (i < j) {
        
        while (i < j && key >= array[j].integerValue) {
            j--;
        }
        [array exchangeObjectAtIndex:i withObjectAtIndex:j];
        
        while (i < j && key <= array[i].integerValue) {
            i++;
        }
        [array exchangeObjectAtIndex:i withObjectAtIndex:j];
    }
    [self sortWithArray:array leftIndex:leftIndex rightIndex:i - 1];
    [self sortWithArray:array leftIndex:i + 1 rightIndex:rightIndex];
    return array;
}

@end
